azurelinuxagent/ga/rsm_version_updater.py

```diff
@@ -0,0 +1,159 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+
+import datetime
+import glob
+import os
+
+from azurelinuxagent.common import conf, logger
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.exception import AgentUpgradeExitException, AgentUpdateError
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.version import CURRENT_VERSION, AGENT_NAME
+from azurelinuxagent.ga.ga_version_updater import GAVersionUpdater, VMDisabledRSMUpdates
+from azurelinuxagent.ga.guestagent import GuestAgent
+
+
+class RSMVersionUpdater(GAVersionUpdater):
+    def __init__(self, gs_id, daemon_version, last_attempted_rsm_version_update_time):
+        super(RSMVersionUpdater, self).__init__(gs_id)
+        self._daemon_version = daemon_version
+        self._last_attempted_rsm_version_update_time = last_attempted_rsm_version_update_time
+
+    @staticmethod
+    def _get_all_agents_on_disk():
+        path = os.path.join(conf.get_lib_dir(), "{0}-*".format(AGENT_NAME))
+        return [GuestAgent.from_installed_agent(path=agent_dir) for agent_dir in glob.iglob(path) if
+                os.path.isdir(agent_dir)]
+
+    def _get_available_agents_on_disk(self):
+        available_agents = [agent for agent in self._get_all_agents_on_disk() if agent.is_available]
+        return sorted(available_agents, key=lambda agent: agent.version, reverse=True)
+
+    def is_update_allowed_this_time(self):
+        """
+        update is allowed once per (as specified in the conf.get_autoupdate_frequency())
+        If update allowed, we update the last_attempted_rsm_version_update_time to current time.
+        """
+        now = datetime.datetime.now()
+
+        if self._last_attempted_rsm_version_update_time != datetime.datetime.min:
+            next_attempt_time = self._last_attempted_rsm_version_update_time + datetime.timedelta(
+                seconds=conf.get_agentupdate_frequency())
+        else:
+            next_attempt_time = now
+
+        if next_attempt_time > now:
+            return False
+        self._last_attempted_rsm_version_update_time = now
```

Comment:
Why do we update the last_attempted_rsm_version_update_time before we actually proceed with the update? If for some reason we can't update to the provided version, then we would have to wait one whole frequency before attempting another update with a different version. 

Not necessarily suggesting we change this, but just wanting to understand the decision

Reply:
This has been pointed out at least twice in previous iterations of this PR. I agree with @maddieford's comment